src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -499,352 +527,305 @@ public Task<Compilation> GetCompilationAsync(SolutionCompilationState compilatio
+                            // Include the tree even if the content of the document failed to load.
+                            trees.Add(await documentState.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false));
                         }
-                        catch (Exception e) when (FatalError.ReportAndPropagateUnlessCanceled(e, cancellationToken, ErrorSeverity.Critical))
-                        {
-                            throw ExceptionUtilities.Unreachable();
-                        }
-                    }
-
-                    [PerformanceSensitive(
-                        "https://github.com/dotnet/roslyn/issues/23582",
-                        Constraint = "Avoid calling " + nameof(Compilation.AddSyntaxTrees) + " in a loop due to allocation overhead.")]
-                    async Task<AllSyntaxTreesParsedState> BuildAllSyntaxTreesParsedStateFromScratchAsync(
-                        NoCompilationState noCompilationState,
-                        CancellationToken cancellationToken)
-                    {
-                        try
-                        {
-                            var compilation = CreateEmptyCompilation();
 
-                            using var _ = ArrayBuilder<SyntaxTree>.GetInstance(ProjectState.DocumentStates.Count, out var trees);
-                            foreach (var documentState in ProjectState.DocumentStates.GetStatesInCompilationOrder())
-                            {
-                                cancellationToken.ThrowIfCancellationRequested();
-                                // Include the tree even if the content of the document failed to load.
-                                trees.Add(await documentState.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false));
-                            }
+                        compilation = compilation.AddSyntaxTrees(trees);
 
-                            compilation = compilation.AddSyntaxTrees(trees);
-                            var allSyntaxTreesParsedState = new AllSyntaxTreesParsedState(compilation, noCompilationState.GeneratorInfo, compilationWithGeneratedDocuments: null);
-                            WriteState(allSyntaxTreesParsedState);
-                            return allSyntaxTreesParsedState;
-                        }
-                        catch (Exception e) when (FatalError.ReportAndPropagateUnlessCanceled(e, cancellationToken, ErrorSeverity.Critical))
-                        {
-                            throw ExceptionUtilities.Unreachable();
-                        }
+                        // We only got here when we had no compilation state at all.  So we couldn't have gotten
+                        // here from a frozen state (as a frozen state always ensures we have a
+                        // WithCompilationTrackerState).  As such, we can safely still preserve that we're not
+                        // frozen here.
```

Comment:
I agree, want to assert it?

Reply:
there's nothing we can assert.  the NoCompilationState literally does not have an IsFrozen flag :)